aleph/logic/mapping.py

```diff
@@ -0,0 +1,113 @@
+import logging
+from banal import first
+from flask import request
+from werkzeug.exceptions import BadRequest
+from followthemoney import model
+from banal import hash_data, keys_values
+
+from aleph.core import archive, db
+from aleph.views.util import get_index_entity, get_session_id
+from aleph.logic.aggregator import get_aggregator
+from aleph.logic.collections import refresh_collection
+from aleph.queues import queue_task, OP_INDEX, OP_BULKDELETE, OP_BULKLOAD
+from aleph.model import Mapping
+
+log = logging.getLogger(__name__)
+
+
+def get_mapping_query(mapping):
+    table = get_index_entity(mapping.table_id, request.authz.READ)
+    properties = table.get('properties', {})
+    csv_hash = first(properties.get('csvHash'))
+    query = {
+        'entities': mapping.query,
+        'proof': mapping.table_id,
+    }
+    if csv_hash:
+        url = archive.generate_url(csv_hash)
+        if not url:
+            local_path = archive.load_file(csv_hash)
+            if local_path is not None:
+                url = local_path.as_posix()
+        if url is not None:
+            query['csv_url'] = url
+            return {
+                'query': query,
+                'mapping_id': mapping.id,
+            }
+        raise BadRequest("Could not generate csv url for the table")
+    raise BadRequest("Source table doesn't have a csvHash")
+
+
+def bulk_load(stage, collection, config):
+    """Bulk load entities from a CSV file or SQL database.
+
+    This is done by mapping the rows in the source data to entities and links
+    which can be understood by the entity index.
+    """
+    queries = keys_values(config, 'queries', 'query')
+    mapping_id = config.get('mapping_id')
+    mapping = Mapping.by_id(mapping_id)
+    config['entity_ids'] = []
+    for query in queries:
+        try:
+            entity_ids = load_mapping_query(stage, collection, hash_data(query), query)  # noqa
+            config['entity_ids'].extend(entity_ids)
+        except Exception as exc:
+            if mapping:
```

Comment:
What is our strategy here for removing the old "bulkload" - are we doing it or not? Seems like the API is gone but the CLI command and logic layer support is still there...

Reply:
I think the CLI command is still handy for loading mappings files directly and we should keep it. We're using it in tests in a couple of places to quickly load entities without creating a Mapping object which is also somewhat useful.